### DISCUSSION TITLE
Ancestor foreign keys are not enforced in Oracle DB when inserting data

### DISCUSSION THREAD

**Discussion by bart-kra** [sqlalchemy/sqlalchemy#12210] [Usage Questions] [answered]
Hello,

I am a newbie to sqlalchemy and met a problem with ancestor foreign keys while inserting sqlalchemy object to Oracle database. For simplicity the data model has been created as follows:

```
@dataclass
class Base(DeclarativeBase):
    metadata = metadata

@dataclass
class AncestorTable(Base):
    __tablename__ = "tb_ancestor"
    id: Mapped[int] = mapped_column(primary_key=True)
    ancestor_value: Mapped[float]

    parent: Mapped["ParentTable"] = relationship()

    def __init__(self):
        self.parent = ParentTable()

@dataclass
class ParentTable(Base):
    __tablename__ = "tb_parent"
    id: Mapped[int] = mapped_column(primary_key=True)
    ancestor_id: Mapped[int] = mapped_column(ForeignKey("tb_ancestor.id"))
    parent_value: Mapped[float]

    children: Mapped[List["Child"]] = relationship()

@dataclass
class Child(Base):
    __tablename__ = "tb_children"
    id: Mapped[int] = mapped_column(primary_key=True)
    ancestor_id: Mapped[int] = mapped_column(ForeignKey("tb_ancestor.id"))
    parent_id: Mapped[int] = mapped_column(ForeignKey("tb_parent.id"))
    child_value: Mapped[float]
```

Then some dummy data has been assigned:
```
rs = AncestorTable()

rs.ancestor_value = 1
rs.parent.parent_value = 2
rs.parent.children.append(Child(child_value=3))
rs.parent.children.append(Child(child_value=4))
```

These objects are being inserted to Oracle tables with auto-increment id columns:

```
with Session(engine) as session:
    session.add_all([rs])
    session.commit()
```

The process of inserting data into the tb_ancestor and tb_parent tables is functioning correctly, with IDs being generated as expected. Additionally, the ancestor_id foreign key in the parent table is properly enforced. However, an issue arises with the tb_children table: the ancestor_id foreign key constraint is not being enforced. 
Overall, while the foreign key constraints for parent relationships are consistently applied, those for ancestor relationships are not.

Problematic tb_children table:
ID | ANCESTOR_ID | PARENT_ID | CHILD_VALUE
---|-----------------|------------- |----------------
 1 |        (null)         |         1         |           3
 2 |        (null)         |         1         |           4

Any help would be greatly appreciated! 
Thank you, Bart

**Comment by CaselIT** [comment]
Hi,

There is nothing telling sqlalchemy how it should automatically populate the fk.
You may be able to do that by adding a relationship between Child and Ancestor

**Reply by bart-kra** [reply]
Thank you for your reply.
I am following your suggestion and have tried two approaches:
1) In AncestorTable class I have added:
```
children: Mapped[List['Child']] = relationship()
```
But it did not work.

---
2) In Child class I have added:
```
ancestor = relationship('AncestorTable')
```
It did not solove my problem as well.

---
So I have tried to add both lines together and then the code throws a warning:
```
relationship 'Child.ancestor' will copy column tb_ancestor.ID to column tb_children.ancestor_id, which conflicts with relationship(s): 'AncestorTable.children' (copies column tb_ancestor.ID to tb_children.ancestor_id) (...)
```
As I understand correctly above warning confirms that solution 1) or 2) should have the same effect and, in conclusion, solove my problem. However... It did not. ANCESTOR_ID in tb_children is still now enforced.

Any advices what did I do wrong?

**Reply by CaselIT** [reply]
I think the issue is that you are trying to attach two different relationships to the same object and sqlalchemy doesn't like that.

I'm not sure if in this case there is a way of automating that. @zzzeek do you have a suggestion of this use case?

**Comment by zzzeek** [comment, marked as answer]
you effectively have a composite foreign key from Child(parent_id, ancestor_id) to Parent(id, ancestor_id), which you can't set up in Oracle server since Parent.ancestor_id is not part of the primary key, however you can achieve this with [primaryjoin](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#relationship-primaryjoin) where by using the [foreign()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.foreign) modifier you tell the ORM what columns would be populated from where

```py
    children: Mapped[List["Child"]] = relationship(
        primaryjoin=lambda: and_(
            foreign(Child.parent_id) == ParentTable.id,
            foreign(Child.ancestor_id) == ParentTable.ancestor_id,
        )
    )

```

POC

```py
from __future__ import annotations

from typing import List

from sqlalchemy import and_
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Identity
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import foreign
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class AncestorTable(Base):
    __tablename__ = "tb_ancestor"
    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    ancestor_value: Mapped[float]

    parent: Mapped["ParentTable"] = relationship()

    def __init__(self):
        self.parent = ParentTable()


class ParentTable(Base):
    __tablename__ = "tb_parent"
    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    ancestor_id: Mapped[int] = mapped_column(ForeignKey("tb_ancestor.id"))
    parent_value: Mapped[float]

    children: Mapped[List["Child"]] = relationship(
        primaryjoin=lambda: and_(
            foreign(Child.parent_id) == ParentTable.id,
            foreign(Child.ancestor_id) == ParentTable.ancestor_id,
        )
    )


class Child(Base):
    __tablename__ = "tb_children"
    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    ancestor_id: Mapped[int] = mapped_column(ForeignKey("tb_ancestor.id"))
    parent_id: Mapped[int] = mapped_column(ForeignKey("tb_parent.id"))
    child_value: Mapped[float]


rs = AncestorTable()

rs.ancestor_value = 1
rs.parent.parent_value = 2
rs.parent.children.append(Child(child_value=3))
rs.parent.children.append(Child(child_value=4))

engine = create_engine("oracle+oracledb://scott:tiger@oracle18c/xe", echo=True)
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as session:
    session.add_all([rs])
    session.commit()
```

**Comment by bart-kra** [comment]
I  do sincerely appreciate your support. @zzzeek, your POC is clearly what I was looking for and it has soloved my problem. That solution is not very straightforward in more complicated data models; however, it still fulfils its task.
Thank you for your engagement, guys.
